dpd/src/mcast/mod.rs

```diff
@@ -1333,15 +1407,19 @@ fn ensure_external_group_exists(
     group_ip: IpAddr,
     members: &HashSet<MulticastGroupMember>,
     group_entry: &mut MulticastGroup,
-    mcast: &mut MulticastGroupData,
+    pre_allocated_id: Option<ScopedGroupId>,
 ) -> DpdResult<()> {
     if group_entry.external_group_id.is_none()
         && members.iter().any(|m| m.direction == Direction::External)
     {
-        let group_id = mcast.generate_group_id()?;
-        create_asic_group(s, group_id, group_ip)?;
+        let scoped_group_id = pre_allocated_id.ok_or_else(|| {
+            DpdError::Invalid(
+                "external group ID should have been pre-allocated".to_string(),
+            )
```

Comment:
This feels like an invariant maintained server-side, rather than something derived from poor user input. `DpdError::Invalid` will convert into an HTTP 400, whereas this seems like a 500 on that note. `DpdError::Other` might fill that niche?